src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,175 @@
+        const expression = parent && isValidExpressionOrStatement(parent) ? getExpression(parent) : undefined;
+        if (!expression) return undefined;
+
+        const checker = program.getTypeChecker();
+
+        if (isBinaryExpression(expression)) {
+            const lastPropertyAccessChain = getLastPropertyAccessChain(expression.right);
+            if (!lastPropertyAccessChain) return undefined;
+            const firstOccurrence = getFirstOccurrence(expression, lastPropertyAccessChain, checker);
+            return firstOccurrence ? { lastPropertyAccessChain, firstOccurrence, expression } : undefined;
+        }
+
+        if (isConditionalExpression(expression)) {
+            const whenTrue = expression.whenTrue;
+            const condition = expression.condition;
+            if ((isIdentifier(condition) || isPropertyAccessExpression(condition)) &&
+                isPropertyAccessExpression(whenTrue) && checker.isMatchingReference(whenTrue.expression, condition)) {
+                // The ternary expression and nullish coalescing would result in different return values if c is nullish so do not offer a refactor
+                const type = checker.getTypeAtLocation(whenTrue.name);
+                return !checker.isNullableType(type) ? { lastPropertyAccessChain: whenTrue, firstOccurrence: condition, expression } : undefined;
+            }
+        }
+        return undefined;
+    }
+
+    function getFirstOccurrence(expression: BinaryExpression, lastPropertyAccessChain: PropertyAccessExpression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
+        if (expression.operatorToken.kind !== SyntaxKind.AmpersandAmpersandToken && expression.operatorToken.pos <= lastPropertyAccessChain.pos) return undefined;
+        // ensure that each sequential operand in range matches the longest access chain
+        let checkNode = expression.left;
+        let firstOccurrence: PropertyAccessExpression | Identifier = lastPropertyAccessChain;
+        while (isBinaryExpression(checkNode) && checkNode.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken &&
+            (isPropertyAccessExpression(checkNode.right) || isIdentifier(checkNode.right))) {
+            const previousReference = isPropertyAccessExpression(firstOccurrence) ? firstOccurrence.expression : firstOccurrence;
+            if (!checker.isMatchingReference(previousReference, checkNode.right)) {
+                return undefined;
+            }
+            firstOccurrence = checkNode.right;
+            checkNode = checkNode.left;
+        }
+        // check final LHS node
+        if (isIdentifier(checkNode) || isPropertyAccessExpression(checkNode)) {
+            const previousReference = isPropertyAccessExpression(firstOccurrence) ? firstOccurrence.expression : firstOccurrence;
+            if (isPropertyAccessExpression(firstOccurrence) && checker.isMatchingReference(previousReference, checkNode)) {

```

Comment:
since you don't bail here, it seems like this should support suffix sequences like

`a && b && c && c.d --> a && b && c?.d`

Are there tests like this?

Reply:
These are all valid and tested.

```ts
foo && bar && [|a && a.b|]

[|a && a.b|] && foo && bar

[|a && a.b|] && [|x && x.y|]
```